File: next/plugins/inspector/inspector.py
"""viewer.bus-inspector — debug plugin capturing all bus traffic (framework A.10).

Subscribes to `>` (fully open subscription, no privilege), filters its own
origin client-side, keeps a bounded ring buffer, streams matching frames on
`bus-inspector:_:matches`, and exposes pause/resume/clear/set-filter plus an
explicit paginated snapshot RPC (framework section 5.6 history rule). Emits a
`bus-inspector:_:stats` mailbox with rate/dropped counters; downsamples the
match stream when traffic exceeds the emit threshold.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import logging
import os
import signal
import time
from collections import deque
from typing import Any

from sdk import Ctx, Plugin, match, slot

logger = logging.getLogger(__name__)

MANIFEST: dict[str, Any] = {
    "id": "bus-inspector",
    "version": "0.1.0",
    "slots": {"set-filter": {}, "pause": {}, "resume": {}, "clear": {}, "snapshot": {}},
    "emits": {"matches": {}, "stats": {}},
}

MATCHES_CHANNEL = "bus-inspector:_:matches"
STATS_CHANNEL = "bus-inspector:_:stats"

# Snapshot replies are byte-budgeted (serialized JSON, newest-first) so they
# can never exceed the kernel's 1 MiB frame limit — an oversized reply is
# rejected by the kernel and the RPC caller hangs until timeout.
SNAPSHOT_BUDGET = 800_000


class BusInspectorPlugin(Plugin):
    manifest = MANIFEST

    def __init__(
        self,
        *,
        ring_size: int = 5000,
        emit_threshold: int = 200,
        stats_interval: float = 1.0,
        echo: bool = False,
        **client_kwargs: Any,
    ) -> None:
        super().__init__(**client_kwargs)
        self.ring: deque[dict[str, Any]] = deque(maxlen=ring_size)
        self.seq = 0
        self.filter: dict[str, Any] = {}
        self.paused = False
        self.echo = echo
        self.emit_threshold = emit_threshold
        self.stats_interval = stats_interval
        self.captured = 0
        self.emitted = 0
        self.dropped = 0
        self._window_start = time.monotonic()
        self._window_emitted = 0
        self._rate = 0.0
        self._window_captured = 0
        self._stats_task: asyncio.Task[None] | None = None

    # -------------------------------------------------------------- lifecycle

    async def on_start(self) -> None:
        self._stats_task = asyncio.create_task(self._stats_loop())

    async def on_stop(self) -> None:
        if self._stats_task is not None:
            self._stats_task.cancel()

    # ---------------------------------------------------------------- capture

    @slot(">")
    async def capture(self, ctx: Ctx) -> None:
        origin = ctx.origin or {}
        # Echo prevention is the consumer's job (framework section 5.3).
        if origin.get("plugin") == self.manifest["id"]:
            return
        self.seq += 1
        entry = {
            "seq": self.seq,
            "ts": ctx.ts,
            "type": ctx.frame.get("type"),
            "channel": ctx.channel,
            "origin": origin,
            "trace_id": ctx.trace_id,
            "depth": ctx.depth,
            "value": ctx.value,
        }
        self.ring.append(entry)
        self.captured += 1
        self._window_captured += 1
        if self.echo:
            rendered = json.dumps(entry["value"], separators=(",", ":"), ensure_ascii=False)
            print(f"{entry['ts']} {entry['type']:>8} {entry['channel']} "
                  f"{origin.get('plugin', '?')} {rendered[:200]}", flush=True)
        if self.paused or not self._matches(entry):
            return
        now = time.monotonic()
        if now - self._window_start >= 1.0:
            self._window_start = now
            self._window_emitted = 0
        if self._window_emitted >= self.emit_threshold:
            self.dropped += 1
            return
        self._window_emitted += 1
        self.emitted += 1
        # The match stream is a fresh observation channel, NOT a causal
        # continuation of the captured frame — otherwise capturing a depth-7
        # frame would emit at depth 8 and hit the kernel's loop guard.
        assert self.client is not None
        await self.client.publish(MATCHES_CHANNEL, entry)

    def _matches(self, entry: dict[str, Any]) -> bool:
        f = self.filter
        if not f:
            return True
        if f.get("channel") and not match(f["channel"], entry["channel"]):
            return False
        if f.get("type") and entry["type"] != f["type"]:
            return False
        if f.get("origin") and entry["origin"].get("plugin") != f["origin"]:
            return False
        if f.get("trace_id") and entry.get("trace_id") != f["trace_id"]:
            return False
        if f.get("text"):
            haystack = json.dumps(entry["value"], ensure_ascii=False)
            if f["text"] not in haystack:
                return False
        return True

    # ------------------------------------------------------------------ slots

    @slot("bus-inspector:_:set-filter")
    async def set_filter(self, ctx: Ctx) -> None:
        value = ctx.value if isinstance(ctx.value, dict) else {}
        allowed = {"channel", "type", "origin", "trace_id", "text"}
        self.filter = {key: value[key] for key in allowed if value.get(key)}
        await self._publish_stats()
        await ctx.respond({"filter": self.filter})

    @slot("bus-inspector:_:pause")
    async def pause(self, ctx: Ctx) -> None:
        self.paused = True
        await self._publish_stats()
        await ctx.respond({"paused": True})

    @slot("bus-inspector:_:resume")
    async def resume(self, ctx: Ctx) -> None:
        self.paused = False
        await self._publish_stats()
        await ctx.respond({"paused": False})

    @slot("bus-inspector:_:clear")
    async def clear(self, ctx: Ctx) -> None:
        self.ring.clear()
        await ctx.respond({"cleared": True})

    @slot("bus-inspector:_:snapshot")
    async def snapshot(self, ctx: Ctx) -> None:
        """Explicit paginated history (framework section 5.6): ``limit`` +
        ``before_seq`` cursor; never an implicit read of the live stream."""

        value = ctx.value if isinstance(ctx.value, dict) else {}
        limit = int(value.get("limit", 100))
        before_seq = value.get("before_seq")
        # Newest-first under a serialized-size budget; keep at least the
        # newest matching entry regardless of size.
        entries: list[dict[str, Any]] = []
        budget = SNAPSHOT_BUDGET
        for entry in reversed(self.ring):
            if len(entries) >= limit:
                break
            if before_seq is not None and entry["seq"] >= before_seq:
                continue
            size = len(json.dumps(entry))
            if entries and size > budget:
                break
            entries.append(entry)
            budget -= size
        entries.reverse()
        await ctx.respond({"entries": entries})

    # ------------------------------------------------------------------ stats

    async def _stats_loop(self) -> None:
        while True:
            await asyncio.sleep(self.stats_interval)
            self._rate = self._window_captured / self.stats_interval
            self._window_captured = 0
            await self._publish_stats()

    async def _publish_stats(self) -> None:
        if self.client is None or not self.client.connected:
            return
        await self.client.set(
            STATS_CHANNEL,
            {
                "captured": self.captured,
                "emitted": self.emitted,
                "dropped": self.dropped,
                "rate_per_sec": self._rate,
                "paused": self.paused,
                "filter": self.filter,
                "ring_size": self.ring.maxlen,
                "ring_used": len(self.ring),
            },
        )

    # ------------------------------------------------------------- entrypoint

    def run(self) -> None:
        parser = argparse.ArgumentParser(description="Viewer bus-inspector debug plugin")
        parser.add_argument("--kernel-ws", required=True)
        parser.add_argument("--ring-size", type=int, default=5000)
        parser.add_argument("--echo", action="store_true", help="print captured frames to stdout")
        args = parser.parse_args()
        logging.basicConfig(
            level=logging.INFO,
            format="%(asctime)s %(levelname)s %(name)s: %(message)s",
        )
        self.ring = deque(maxlen=args.ring_size)
        self.echo = args.echo
        managed = os.environ.get("VIEWER_MANAGED") == "1"

        async def main() -> None:
            await self.start(args.kernel_ws, managed=managed)
            stopped = asyncio.Event()
            loop = asyncio.get_running_loop()
            for signum in (signal.SIGINT, signal.SIGTERM):
                loop.add_signal_handler(signum, stopped.set)
            await stopped.wait()
            await self.stop()

        asyncio.run(main())

File: next-go/internal/plugins/inspector/inspector.go
// Package inspector implements the bus-inspector debugging plugin.
package inspector

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"strings"
	"sync"
	"time"

	"viewer/internal/busclient"
	"viewer/internal/protocol"
)

const (
	MatchesChannel = "bus-inspector:_:matches"
	StatsChannel   = "bus-inspector:_:stats"
	SnapshotBudget = 800_000
)

var Manifest = busclient.Manifest{
	ID: "bus-inspector", Version: "0.1.0",
	Slots: map[string]any{
		"set-filter": map[string]any{}, "pause": map[string]any{},
		"resume": map[string]any{}, "clear": map[string]any{}, "snapshot": map[string]any{},
	},
	Emits: map[string]any{"matches": map[string]any{}, "stats": map[string]any{}},
}

type Config struct {
	KernelWS      string
	RingSize      int
	Echo          bool
	EmitThreshold int
	StatsInterval time.Duration
}

func DefaultConfig() Config {
	return Config{RingSize: 5000, EmitThreshold: 200, StatsInterval: time.Second}
}

type Entry struct {
	Seq     uint64            `json:"seq"`
	TS      int64             `json:"ts"`
	Type    string            `json:"type"`
	Channel string            `json:"channel"`
	Origin  map[string]string `json:"origin"`
	TraceID string            `json:"trace_id"`
	Depth   int               `json:"depth"`
	Value   any               `json:"value"`
}

type Stats struct {
	Captured   uint64            `json:"captured"`
	Emitted    uint64            `json:"emitted"`
	Dropped    uint64            `json:"dropped"`
	RatePerSec float64           `json:"rate_per_sec"`
	Paused     bool              `json:"paused"`
	Filter     map[string]string `json:"filter"`
	RingSize   int               `json:"ring_size"`
	RingUsed   int               `json:"ring_used"`
}

type Plugin struct {
	config                     Config
	client                     *busclient.Client
	mu                         sync.Mutex
	ring                       []Entry
	seq                        uint64
	filter                     map[string]string
	paused                     bool
	captured, emitted, dropped uint64
	windowCaptured             uint64
	rate                       float64
	windowStart                time.Time
	emitWindowStart            time.Time
	windowEmitted              int
}

func New(config Config) (*Plugin, error) {
	defaults := DefaultConfig()
	if config.KernelWS == "" {
		return nil, fmt.Errorf("kernel websocket is required")
	}
	if config.RingSize == 0 {
		config.RingSize = defaults.RingSize
	}
	if config.RingSize < 1 {
		return nil, fmt.Errorf("ring size must be positive")
	}
	if config.EmitThreshold <= 0 {
		config.EmitThreshold = defaults.EmitThreshold
	}
	if config.StatsInterval <= 0 {
		config.StatsInterval = defaults.StatsInterval
	}
	now := time.Now()
	return &Plugin{config: config, filter: make(map[string]string), windowStart: now, emitWindowStart: now}, nil
}

func (p *Plugin) Run(ctx context.Context) error {
	managed := os.Getenv("VIEWER_MANAGED") == "1"
	p.client = busclient.New(p.config.KernelWS, Manifest, busclient.WithManaged(managed))
	if _, err := p.client.Subscribe(">", p.capture); err != nil {
		return err
	}
	handlers := map[string]func(busclient.Frame){
		"set-filter": p.setFilter, "pause": p.pause, "resume": p.resume,
		"clear": p.clear, "snapshot": p.snapshot,
	}
	for name, handler := range handlers {
		if _, err := p.client.Subscribe("bus-inspector:_:"+name, handler); err != nil {
			return err
		}
	}
	if err := p.client.Connect(ctx); err != nil {
		return fmt.Errorf("connect bus-inspector: %w", err)
	}
	p.publishStats()
	ticker := time.NewTicker(p.config.StatsInterval)
	defer ticker.Stop()
	for {
		select {
		case now := <-ticker.C:
			p.mu.Lock()
			elapsed := now.Sub(p.windowStart).Seconds()
			if elapsed > 0 {
				p.rate = float64(p.windowCaptured) / elapsed
			}
			p.windowCaptured, p.windowStart = 0, now
			p.mu.Unlock()
			p.publishStats()
		case <-ctx.Done():
			return p.client.Close()
		}
	}
}

func (p *Plugin) capture(frame busclient.Frame) {
	if frame.Origin != nil && frame.Origin.Plugin == Manifest.ID {
		return
	}
	origin := map[string]string{}
	if frame.Origin != nil {
		origin["plugin"], origin["instance"] = frame.Origin.Plugin, frame.Origin.Instance
	}
	p.mu.Lock()
	p.seq++
	entry := Entry{Seq: p.seq, TS: frame.TS, Type: frame.Type, Channel: frame.Channel,
		Origin: origin, TraceID: frame.TraceID, Depth: frame.Depth, Value: frame.Value}
	if len(p.ring) == p.config.RingSize {
		copy(p.ring, p.ring[1:])
		p.ring[len(p.ring)-1] = entry
	} else {
		p.ring = append(p.ring, entry)
	}
	p.captured++
	p.windowCaptured++
	paused, matches := p.paused, p.matchesLocked(entry)
	now := time.Now()
	shouldEmit := !paused && matches
	if shouldEmit {
		if now.Sub(p.emitWindowStart) >= time.Second {
			p.emitWindowStart, p.windowEmitted = now, 0
		}
		if p.windowEmitted >= p.config.EmitThreshold {
			p.dropped++
			shouldEmit = false
		} else {
			p.windowEmitted++
			p.emitted++
		}
	}
	echo := p.config.Echo
	p.mu.Unlock()
	if echo {
		value, _ := json.Marshal(entry.Value)
		if len(value) > 200 {
			value = value[:200]
		}
		fmt.Printf("%d %8s %s %s %s\n", entry.TS, entry.Type, entry.Channel, entry.Origin["plugin"], value)
	}
	if shouldEmit && p.client != nil {
		if err := p.client.Publish(context.Background(), MatchesChannel, entry); err != nil {
			slog.Warn("publish inspector match", "error", err)
		}
	}
}

func (p *Plugin) matchesLocked(entry Entry) bool {
	f := p.filter
	if pattern := f["channel"]; pattern != "" && !protocol.ChannelMatches(pattern, entry.Channel) {
		return false
	}
	if value := f["type"]; value != "" && value != entry.Type {
		return false
	}
	if value := f["origin"]; value != "" && value != entry.Origin["plugin"] {
		return false
	}
	if value := f["trace_id"]; value != "" && value != entry.TraceID {
		return false
	}
	if text := f["text"]; text != "" {
		encoded, _ := json.Marshal(entry.Value)
		if !strings.Contains(string(encoded), text) {
			return false
		}
	}
	return true
}

func (p *Plugin) setFilter(frame busclient.Frame) {
	request, _ := frame.Value.(map[string]any)
	if request["_cancel"] == true {
		return
	}
	filter := make(map[string]string)
	for _, key := range []string{"channel", "type", "origin", "trace_id", "text"} {
		if value, ok := request[key].(string); ok && value != "" {
			filter[key] = value
		}
	}
	p.mu.Lock()
	p.filter = filter
	p.mu.Unlock()
	p.publishStats()
	p.respond(request, map[string]any{"filter": filter})
}

func (p *Plugin) pause(frame busclient.Frame) {
	request, _ := frame.Value.(map[string]any)
	if request["_cancel"] == true {
		return
	}
	p.mu.Lock()
	p.paused = true
	p.mu.Unlock()
	p.publishStats()
	p.respond(request, map[string]any{"paused": true})
}

func (p *Plugin) resume(frame busclient.Frame) {
	request, _ := frame.Value.(map[string]any)
	if request["_cancel"] == true {
		return
	}
	p.mu.Lock()
	p.paused = false
	p.mu.Unlock()
	p.publishStats()
	p.respond(request, map[string]any{"paused": false})
}

func (p *Plugin) clear(frame busclient.Frame) {
	request, _ := frame.Value.(map[string]any)
	if request["_cancel"] == true {
		return
	}
	p.mu.Lock()
	p.ring = nil
	p.mu.Unlock()
	p.publishStats()
	p.respond(request, map[string]any{"cleared": true})
}

func (p *Plugin) snapshot(frame busclient.Frame) {
	request, _ := frame.Value.(map[string]any)
	if request["_cancel"] == true {
		return
	}
	limit := 100
	if raw, ok := request["limit"].(float64); ok {
		limit = int(raw)
	}
	if limit < 1 {
		limit = 1
	}
	var before uint64
	if raw, ok := request["before_seq"].(float64); ok && raw > 0 {
		before = uint64(raw)
	}
	p.mu.Lock()
	ring := append([]Entry(nil), p.ring...)
	p.mu.Unlock()
	p.respond(request, map[string]any{"entries": snapshotEntries(ring, limit, before)})
}

func snapshotEntries(ring []Entry, limit int, before uint64) []Entry {
	entries := make([]Entry, 0, limit)
	used := 0
	for i := len(ring) - 1; i >= 0 && len(entries) < limit; i-- {
		entry := ring[i]
		if before != 0 && entry.Seq >= before {
			continue
		}
		encoded, _ := json.Marshal(entry)
		size := len(encoded)
		if len(entries) > 0 {
			size++
		}
		if len(entries) > 0 && used+size > SnapshotBudget {
			break
		}
		entries = append(entries, entry)
		used += size
	}
	return entries
}

func (p *Plugin) publishStats() {
	if p.client == nil || !p.client.Connected() {
		return
	}
	p.mu.Lock()
	filter := make(map[string]string, len(p.filter))
	for key, value := range p.filter {
		filter[key] = value
	}
	stats := Stats{Captured: p.captured, Emitted: p.emitted, Dropped: p.dropped,
		RatePerSec: p.rate, Paused: p.paused, Filter: filter,
		RingSize: p.config.RingSize, RingUsed: len(p.ring)}
	p.mu.Unlock()
	if err := p.client.Set(context.Background(), StatsChannel, stats); err != nil {
		slog.Warn("publish inspector stats", "error", err)
	}
}

func (p *Plugin) respond(request map[string]any, result any) {
	reply, replyOK := request["_reply_to"].(string)
	corr, corrOK := request["_corr"].(string)
	if !replyOK || !corrOK {
		return
	}
	_ = p.client.Publish(context.Background(), reply, map[string]any{"_corr": corr, "ok": true, "result": result})
}
